core/server/workspace.go

```diff
@@ -641,6 +642,87 @@ func (ws *Workspace) RestartComponent(ctx context.Context, input *api.RestartCom
 	return &api.RestartComponentOutput{}, nil
 }
 
+func (ws *Workspace) GetComponentStatus(ctx context.Context, input *api.GetComponentStatusInput) (*api.GetComponentStatusOutput, error) {
+	id, err := ws.resolveRef(ctx, input.Ref)
+	if err != nil {
+		return nil, fmt.Errorf("resolving ref: %w", err)
+	}
+
+	describeComponentsOutput, err := ws.DescribeComponents(ctx, &api.DescribeComponentsInput{})
+	if err != nil {
+		return nil, err
+	}
+
+	state := process.State{}
+	for _, component := range describeComponentsOutput.Components {
+		if component.ID == id {
+			if err := jsonutil.UnmarshalString(component.State, &state); err != nil {
+				return nil, fmt.Errorf("unmarshalling state: %w", err)
+			}
+
+			status := api.ComponentStatus{
+				ComponentID: id,
+				EnvVars:     state.FullEnvironment,
+			}
+
+			proc, err := psprocess.NewProcess(int32(state.Pid))
```

Comment:
All this stuff that is analyzing unix processes etc should move to the `/providers/unix/components/process` package. If you do that, then you can use the `newController` function to spec and state unmarshaling automatically.

Reply:
To be clear, I think this stuff belongs: https://github.com/deref/exo/blob/fc97c78c4ae0dae695519ab6105551b1dbf9a071/providers/unix/components/process/lifecycle.go#L42

See the above comment about "refresh" too.

Reply:
I understand moving it to `refresh`, which seems reasonable, but I'm not sure I get the "If you do that, then you can use the newController function to spec and state unmarshaling automatically."

Reply:
Sorry, `newController` is a lower-level bit. The really interesting function is:

```go
func (ws *Workspace) control(ctx context.Context, desc api.ComponentDescription, f interface{}) error {
```

There are numerous examples of using it. What it does is instantiate a controller structure, hydrate it with spec and state, execute a method on it, and then save the modified state back to underlying storage.